oodeel/methods/gram.py

```diff
@@ -0,0 +1,295 @@
+        preds = self.op.convert_to_numpy(self.op.argmax(logits, dim=1))
+
+        # We stack the min_maxs for each class depending on the prediction for each
+        # samples
+        min_maxs = []
+        for i in range(len(tensor_stats)):
+            min_maxs.append(self.op.stack([self.min_maxs[label][i] for label in preds]))
+
+        tensor_dev = self._deviation(tensor_stats, min_maxs)
+        score = self.op.mean(
+            self.op.cat(
+                [
+                    self.op.unsqueeze(tensor_dev_l, dim=0) / devnorm_l
+                    for tensor_dev_l, devnorm_l in zip(tensor_dev, self.devnorm)
+                ]
+            ),
+            dim=0,
+        )
+        return self.op.convert_to_numpy(score)
+
+    def _deviation(
+        self, stats: List[TensorType], min_maxs: List[TensorType]
+    ) -> List[TensorType]:
+        """Compute the deviation wrt quantiles (min/max) for feature_maps
+
+        Args:
+            stats (TensorType): The list of gram matrices (stacked power-wise)
+                for which we want to compute the deviation.
+            min_maxs (TensorType): The quantiles (tensorised) to compute the deviation
+                against.
+
+        Returns:
+            List(TensorType): A list with one element per layer containing a tensor of
+                per-sample deviation.
+        """
+        deviation = []
+        for stat, min_max in zip(stats, min_maxs):
+            where_min = self.op.where(stat < min_max[..., 0], 1.0, 0.0)
+            where_max = self.op.where(stat > min_max[..., 1], 1.0, 0.0)
+            deviation_min = (
+                (min_max[..., 0] - stat)
+                / (self.op.abs(min_max[..., 0]) + 1e-6)
+                * where_min
+            )
+            deviation_max = (
+                (stat - min_max[..., 1])
+                / (self.op.abs(min_max[..., 1]) + 1e-6)
+                * where_max
+            )
+            deviation.append(self.op.sum(deviation_min + deviation_max, dim=(1, 2)))
+        return deviation
+
+    def _stat(self, feature_map: TensorType) -> TensorType:
+        """Compute the correlation map (stat) for a given feature map. The values
+        for each power of gram matrix are contained in the same tensor
+
+        Args:
+            feature_map (TensorType): The input feature_map
+
+        Returns:
+            TensorType: The stacked gram matrices power-wise.
+        """
+        fm_s = feature_map.shape
+        stat = []
+        for p in self.orders:
+            feature_map_p = feature_map**p
+            # construct the Gram matrix
+            if len(fm_s) == 2:
+                # prevents from computing a scalar per batch for layers of shape (1,)
+                feature_map_p = self.op.einsum("bi,bj->b", feature_map_p, feature_map_p)
+                feature_map_p = self.op.unsqueeze(feature_map_p, -1)
+                feature_map_p = self.op.unsqueeze(feature_map_p, -1)
```

Comment:
Here the statistic for a dense layer behaves as the L2 norm of the activation which provides little information. Instead, I suggest building a gram matrix analogously to the convolutional case, where the feature matrix of a sample $F_l(D)$ has a shape [dim_layer, 1]. In practice, this seems to give better results for OOD detection from an MLP.

 ```suggestion
            if len(fm_s) == 2:
                # build gram matrix for feature map of shape [dim_dense_layer, 1]
                feature_map_p = self.op.einsum("bi,bj->bij", feature_map_p, feature_map_p)
```